runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -778,23 +778,56 @@ J9::CodeGenerator::lowerTreeIfNeeded(
+         // object, which would require knowing the number of args of the method. To pass that information, we store to
+         // vmThread.floatTemp1 field. The name of the field is misleading, as it is an lconst/iconst being stored. This
+         // is also done for linkToSpecial, as it shares the same bytecode handler as linkToStatic.
          int32_t numParameterStackSlots = node->getSymbol()->castToResolvedMethodSymbol()->getNumParameterSlots();
+         int32_t numArgs = node->getNumArguments();
          TR::Node * numArgsNode = NULL;
-         TR::Node * storeNode = NULL;
+         TR::Node * numArgSlotsNode = NULL;
+         TR::Node * tempSlotStoreNode = NULL;
+         TR::Node * floatTemp1StoreNode = NULL;
          if (self()->comp()->target().is64Bit())
             {
-            numArgsNode = TR::Node::lconst(node, numParameterStackSlots);
-            storeNode = TR::Node::createStore(vmThreadTempSlotSymRef, numArgsNode, TR::lstore);
+            numArgSlotsNode = TR::Node::lconst(node, numParameterStackSlots);
+            tempSlotStoreNode = TR::Node::createStore(self()->comp()->getSymRefTab()->findOrCreateVMThreadTempSlotFieldSymbolRef(),
+                                                      numArgSlotsNode,
+                                                      TR::lstore);
+            if (rm == TR::java_lang_invoke_MethodHandle_linkToStatic || rm ==  TR::java_lang_invoke_MethodHandle_linkToSpecial)
+               {
+               numArgsNode = TR::Node::lconst(node, numArgs);
+               floatTemp1StoreNode = TR::Node::createStore(self()->comp()->getSymRefTab()->findOrCreateVMThreadFloatTemp1SymbolRef(),
+                                                           numArgsNode,
+                                                           TR::lstore);
+               }
             }
          else
             {
-            numArgsNode = TR::Node::iconst(node, numParameterStackSlots);
-            storeNode = TR::Node::createStore(vmThreadTempSlotSymRef, numArgsNode, TR::istore);
+            numArgSlotsNode = TR::Node::iconst(node, numParameterStackSlots);
+            tempSlotStoreNode = TR::Node::createStore(self()->comp()->getSymRefTab()->findOrCreateVMThreadTempSlotFieldSymbolRef(),
+                                                      numArgSlotsNode,
+                                                      TR::istore);
+            if (rm == TR::java_lang_invoke_MethodHandle_linkToStatic || rm ==  TR::java_lang_invoke_MethodHandle_linkToSpecial)
+               {
+               numArgsNode = TR::Node::iconst(node, numArgs);
+               floatTemp1StoreNode = TR::Node::createStore(self()->comp()->getSymRefTab()->findOrCreateVMThreadFloatTemp1SymbolRef(),
+                                                           numArgsNode,
+                                                           TR::istore);
+               }
+            }
```

Comment:
I wonder if the code above could be written with more sharing between the 32/64 bit paths as:
```
bool is64Bit = self()->comp()->target().is64Bit();

if (is64Bit)
   {
   numArgsNode = TR::Node::lconst(node, numParameterStackSlots);
   storeOpCode = TR::lstore;
   numArgSlotsNode = TR::Node::lconst(node, numParameterStackSlots);
   }
else
   {
   numArgsNode = TR::Node::iconst(node, numParameterStackSlots);
   storeOpCode = TR::istore;
   numArgSlotsNode = TR::Node::iconst(node, numParameterStackSlots);
   }

storeNode = TR::Node::createStore(vmThreadTempSlotSymRef, numArgsNode, storeOpCode);

tempSlotStoreNode = TR::Node::createStore(self()->comp()->getSymRefTab()->findOrCreateVMThreadTempSlotFieldSymbolRef(),
                                          numArgSlotsNode,
                                          storeOpCode);

if (rm == TR::java_lang_invoke_MethodHandle_linkToStatic || rm ==  TR::java_lang_invoke_MethodHandle_linkToSpecial)
   {
   numArgsNode = is64Bit ? TR::Node::lconst(node, numArgs) :
                           TR::Node::iconst(node, numArgs);

   floatTemp1StoreNode = TR::Node::createStore(self()->comp()->getSymRefTab()->findOrCreateVMThreadFloatTemp1SymbolRef(),
                                               numArgsNode,
                                               storeOpCode);
   }
```